app/helpers/application_helper.rb

```diff
@@ -68,4 +68,20 @@ def pdf_stylesheet_pack_tag(source)
       wicked_pdf_stylesheet_pack_tag(source)
     end
   end
+
+  def cache_with_locale(key = nil, options = {}, &block)
+    cache(cache_key_with_locale(key, I18n.locale), options) do
+      yield(block)
+    end
+  end
+
+  def cache_key_with_locale(key, locale)
+    Array.wrap(key) + [locale.to_s, locale_digest(locale)]
```

Comment:
But why is `key` optional? I must not be understanding it probably because by my reading, this:

```haml
= cache_with_locale do
  the thing

= cache_with_locale do
  the other thing
```

would cache the first and retrieve with the same key, so wouldn't it yield the same thing?
```
the thing
the thing
```

That doesn't seem right to me. Am I missing a key detail?

Reply:
Another thought, can we automatically protect against this by appending something like this?

```ruby
"#{__FILE__}:#{__LINE__}"
```

Reply:
Good questions! This PR makes more sense if you have a good understanding of Rails's built-in [fragment caching](https://blog.appsignal.com/2018/03/20/fragment-caching-in-rails.html) feature, which would normally be used by calling the method `cache` with some optional arguments and a block. The source code for the `cache` method is also very instructive if you ever want to get into the nuts and bolts of it: https://github.com/rails/rails/blob/main/actionview/lib/action_view/helpers/cache_helper.rb

What `cache_with_locale` does is just call `cache` but with a couple of extra bits of information added to the `key` argument.

On the use of `Array.wrap(key)`; the `key` argument being passed here might be `nil`, or it might be a _single_ object or string, or it might be an array of objects or strings, and we need to add a couple of bits on to it but handle those cases in a certain way. What `Array.wrap` does here is essentially: "if the object is already an array, return it, otherwise wrap it in an array". So `Array.wrap(1)` will return `[1]` and `Array.wrap([1, 2, 3])` will return `[1, 2, 3]`. That's kind of how we need the arguments to end up in the `cache` method after we've appended the extra arguments. It avoids ending up with an array-in-an-array like `[[1, 2], 3, 4]`.

On using `cache_with_locale` without an argument; it's the same interface as using `cache` (which is the built in `ActionView::Helpers::CacheHelpers#cache` method). You're right that using it twice in the same partial without passing an argument would result in the same key being used for both blocks and produce weird results. But it's also perfectly reasonable to use it once per partial without an argument. You can give it an arbitrary string as a name in the case where there are multiple uses:
```ruby
= cache_with_locale "first" do
  the thing

= cache_with_locale "second" do
  the other thing
```
Like this: https://github.com/openfoodfoundation/openfoodnetwork/pull/10772/files#diff-70a3a4c324c26eac717f70ca4ed2c755488f98152a583c3b6298d27113d36cd3R14

The key will already be scoped to the partial, they're not shared across different partials. Here's an example from the logs:
```
Cache read: views/shared/_footer:8c29743a1533118e763c376ae6c09dd2/en/c470d95ed4d2e14b295a05fe3ffa28fb
```

Reply:
> On the use of `Array.wrap(key)`;

Thanks for explaining that. I think that the array splat `*` operator would produce the same effect, but it's probably more of a personal preference. IE:

```
[1] pry(main)> ary = [1, 2, 3]
[3] pry(main)> Array.wrap(ary) + [4, 5]
=> [1, 2, 3, 4, 5]
[4] pry(main)> [*ary, 4, 5]
=> [1, 2, 3, 4, 5]
```

> On using `cache_with_locale` without an argument;

Thanks for linking me back to that article. I did read it first, but didn't notice that it automatically builds the cache key using "a digest of the template tree". I wonder if this contains a unique reference to that line, or just the file? Either way, yes it's fine to use without an argument at least once per file.
So a second time could produce weird results, which is why I was thinking ahead to protect against this. 
But as you say, it's default Rails behaviour. And issues would probably become apparent quite quickly in tests. 
So all good 👍 .